src/x11.cc

```diff
@@ -1205,61 +1084,83 @@ void print_desktop_name(struct text_object *obj, char *p,
+  // Ensures either vertical or horizontal axis are start/end
+  if ((*align & 0b0101) == 0) return;
+
+  Atom strut = ATOM(_NET_WM_STRUT);
+  if (strut != None) {
     /* reserve space at left, right, top, bottom */
-    signed long sizes[12] = {0};
+    uint32_t sizes[*x11_strut::COUNT] = {0};
     int i;
 
-    /* define strut depth */
-    switch (sidenum) {
-      case 0:
-        /* left side */
-        sizes[0] = window.x + window.width;
-        break;
-      case 1:
-        /* right side */
-        sizes[1] = display_width - window.x;
-        break;
-      case 2:
-        /* top side */
-        sizes[2] = window.y + window.height;
-        break;
-      case 3:
-        /* bottom side */
-        sizes[3] = display_height - window.y;
+    switch (vertical_alignment(align)) {
+      case axis_align::START:
+        sizes[*x11_strut::TOP] =
+            std::min(window.y + window.height, display_height);
+        sizes[*x11_strut::TOP_START_X] = window.x;
+        sizes[*x11_strut::TOP_END_X] =
+            std::min(window.x + window.width, display_width);
+      case axis_align::END:
+        sizes[*x11_strut::BOTTOM] =
+            window.y < display_height ? display_height - window.y : 0;
+        sizes[*x11_strut::BOTTOM_START_X] = window.x;
+        sizes[*x11_strut::BOTTOM_END_X] =
+            std::min(window.x + window.width, display_width);
+      case axis_align::MIDDLE:
+        // can't reserve space in middle of the screen
+      default:
         break;
     }
-
-    /* define partial strut length */
-    if (sidenum <= 1) {
-      sizes[4 + (sidenum * 2)] = window.y;
-      sizes[5 + (sidenum * 2)] = window.y + window.height;
-    } else if (sidenum <= 3) {
-      sizes[4 + (sidenum * 2)] = window.x;
-      sizes[5 + (sidenum * 2)] = window.x + window.width;
-    }
-
-    /* check constraints */
-    for (i = 0; i < 12; i++) {
-      if (sizes[i] < 0) {
-        sizes[i] = 0;
-      } else {
-        if (i <= 1 || i >= 8) {
-          if (sizes[i] > display_width) { sizes[i] = display_width; }
-        } else {
-          if (sizes[i] > display_height) { sizes[i] = display_height; }
-        }
-      }
+    // adding `vertical_alignment(align) & 0x1` makes the switch hit MIDDLE if
+    // vertical alignment is set to left or right
+    switch (static_cast<axis_align>(
+        *horizontal_alignment(align) + *vertical_alignment(align) & 0x1)) {
+      case axis_align::START:
+        sizes[*x11_strut::LEFT] =
+            std::min(window.x + window.width, display_width);
+        sizes[*x11_strut::LEFT_START_Y] = window.y;
+        sizes[*x11_strut::LEFT_END_Y] =
+            std::min(window.y + window.height, display_height);
+      case axis_align::END:
+        sizes[*x11_strut::RIGHT] =
+            window.x < display_width ? display_width - window.x : 0;
+        sizes[*x11_strut::RIGHT_START_Y] = window.y;
+        sizes[*x11_strut::RIGHT_END_Y] =
+            std::min(window.y + window.height, display_height);
```

Comment:
In effect, we only have one jump on the value.

I checked with godbolt,
- `window.x < display_width ? display_width - window.x : 0` turns into a saturating sub instruction on achitectures where it's supported.
- `std::min(window.y + window.height, display_height)` doesn't generate jumps either AFAICT, allows the compiler to better optimize the code than clamping in a loop later.

The only downside is that it looks like there's a lot more going on in the final code, but it's just that enum names take up a lot of space. That's IMO better because I couldn't get what `5 + (sidenum * 2)` is mapping to even while looking at variant values.